src/bundle/DependencyInjection/IbexaSolrExtension.php

```diff
@@ -103,18 +103,13 @@ public function getServicePrefix(): string
     }
 
     /**
-     * Loads a specific configuration.
-     *
-     * @param array $configs An array of configuration values
-     * @param \Symfony\Component\DependencyInjection\ContainerBuilder $container A ContainerBuilder instance
-     *
      * @throws \InvalidArgumentException When provided tag is not defined in this extension
-     *
-     * @api
+     * @throws \Exception
      */
-    public function load(array $configs, ContainerBuilder $container)
+    public function load(array $configs, ContainerBuilder $container): void
     {
         $configuration = $this->getConfiguration($configs, $container);
+        assert($configuration !== null);
```

Comment:
Consider using `ConfigurableExtension`. We're not doing anything fancy with configuration processing (customized merging) so we should be good to use it instead.

Reply:
> Consider using `ConfigurableExtension`. We're not doing anything fancy with configuration processing (customized merging) so we should be good to use it instead.

@Steveb-p 

This seems a bit out of the scope of Symfony upgrade. You're asking here to use a different Symfony SPI, which affects the method processing configuration options in a custom way. Can you guarantee, without any QA sanities for semantic configuration, that changing this like that (which I'm assuming you're expecting) is going to be working in the same way?

```diff
diff --git a/src/bundle/DependencyInjection/IbexaSolrExtension.php b/src/bundle/DependencyInjection/IbexaSolrExtension.php
index 5d04213..6acd523 100644
--- a/src/bundle/DependencyInjection/IbexaSolrExtension.php
+++ b/src/bundle/DependencyInjection/IbexaSolrExtension.php
@@ -20,12 +20,12 @@ use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Definition;
 use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
 use Symfony\Component\DependencyInjection\Reference;
-use Symfony\Component\HttpKernel\DependencyInjection\Extension;
+use Symfony\Component\HttpKernel\DependencyInjection\ConfigurableExtension;
 
 /**
  * @phpstan-type SolrHttpClientConfigArray = array{timeout: int, max_retries: int}
  */
-class IbexaSolrExtension extends Extension
+class IbexaSolrExtension extends ConfigurableExtension
 {
     /**
      * Main Solr search handler service ID.
@@ -103,15 +103,12 @@ class IbexaSolrExtension extends Extension
     }
 
     /**
-     * @throws \InvalidArgumentException When provided tag is not defined in this extension
+     * @param array<string, mixed> $mergedConfig
+     *
      * @throws \Exception
      */
-    public function load(array $configs, ContainerBuilder $container): void
+    public function loadInternal(array $mergedConfig, ContainerBuilder $container): void
     {
-        $configuration = $this->getConfiguration($configs, $container);
-        assert($configuration !== null);
-        $config = $this->processConfiguration($configuration, $configs);
-
         // Loading configuration from lib/Resources/config/container
         $loader = new YamlFileLoader(
             $container,
@@ -125,7 +122,7 @@ class IbexaSolrExtension extends Extension
         );
         $loader->load('services.yml');
 
-        $this->processConnectionConfiguration($container, $config);
+        $this->processConnectionConfiguration($container, $mergedConfig);
 
         $container
             ->registerForAutoconfiguration(UpdateSerializerInterface::class)
```

I have no experience with this SPI, so for me it's too risky.

Reply:
@alongosz check what it actually does:
https://github.com/symfony/symfony/blob/b101b71ddacfa664485bb09ec6272971e458f49f/src/Symfony/Component/HttpKernel/DependencyInjection/ConfigurableExtension.php#L35

```php
abstract class ConfigurableExtension extends Extension
{
    /**
     * {@inheritdoc}
     */
    final public function load(array $configs, ContainerBuilder $container)
    {
        $this->loadInternal($this->processConfiguration($this->getConfiguration($configs, $container), $configs), $container);
    }

    /**
     * Configures the passed container according to the merged configuration.
     */
    abstract protected function loadInternal(array $mergedConfig, ContainerBuilder $container);
}
```

It literally does the exact same thing we do. We do it all the time, and everybody does it. Unless you need to customize processing, that's what you do.

It takes all configuration arrays, takes Configuration data from Container and processes it.

`loadInternal` is a protected method.

Reply:
> @alongosz check what it actually does: https://github.com/symfony/symfony/blob/b101b71ddacfa664485bb09ec6272971e458f49f/src/Symfony/Component/HttpKernel/DependencyInjection/ConfigurableExtension.php#L35
> 
> ```
> abstract class ConfigurableExtension extends Extension
> {
>     /**
>      * {@inheritdoc}
>      */
>     final public function load(array $configs, ContainerBuilder $container)
>     {
>         $this->loadInternal($this->processConfiguration($this->getConfiguration($configs, $container), $configs), $container);
>     }
> 
>     /**
>      * Configures the passed container according to the merged configuration.
>      */
>     abstract protected function loadInternal(array $mergedConfig, ContainerBuilder $container);
> }
> ```
> 
> It literally does the exact same thing we do. We do it all the time, and everybody does it. Unless you need to customize processing, that's what you do.
> 
> It takes all configuration arrays, takes Configuration data from Container and processes it.
> 
> `loadInternal` is a protected method.

Ok, switched to `ConfigurableExtension` via [1991874](https://github.com/ibexa/solr/pull/78/commits/1991874b1b1f591531940570bba4179a747e13ac).